File: c/questions/flatten.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* pft(struct TreeNode* node) {
    struct TreeNode* rn = node->right;
    struct TreeNode* tail = node;
    if (node->left) {
        node->right = node->left;
        tail = pft(node->left);
        node->left = NULL;
    }
    if (rn) {
        tail->right = rn;
        tail = pft(rn);
    }
    return tail;
}

void flatten(struct TreeNode* root) {
    if (root) {
        pft(root);
    } else {
        return;
    }
}

File: python/questions/flatten.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        def flate(n: Optional[TreeNode]) -> Optional[TreeNode]:
            if n is None:
                return None
            rn = n.right
            le = flate(n.left)
            if le is None:
                le = n
            n.right = n.left
            n.left = None
            re = flate(rn)
            le.right = rn
            return re if re is not None else le
        flate(root)
